bin/fetch-historic-measurements.js

```diff
@@ -0,0 +1,289 @@
+
+const [nodePath, selfPath, ...args] = process.argv
+if (args.length === 0 || !args[0].startsWith('0x')) {
+  args.unshift(SparkImpactEvaluator.ADDRESS)
+}
+const [contractAddress, startStr, endStr, minerId] = args
+
+const USAGE = `
+Usage:
+  ${nodePath} ${selfPath} [contract-address] range-start range-end [minerId]
+
+Example:
+  ${nodePath} ${selfPath} 2024-10-01T00:00:00Z 2024-10-02T00:00:00Z
+`
+
+if (!contractAddress) {
+  console.error('Missing required argument: contractAddress')
+  console.error(USAGE)
+  process.exit(1)
+}
+
+if (!startStr) {
+  console.error('Missing required argument: range-start')
+  console.error(USAGE)
+  process.exit(1)
+}
+
+const start = new Date(startStr)
+if (Number.isNaN(start.getTime())) {
+  console.error('Invalid range-start: not a valid date-time string')
+  console.error(USAGE)
+  process.exit(1)
+}
+
+if (!endStr) {
+  console.error('Missing required argument: range-start')
+  console.error(USAGE)
+  process.exit(1)
+}
+
+const end = new Date(endStr)
+if (Number.isNaN(end.getTime())) {
+  console.error('Invalid range-end: not a valid date-time string')
+  console.error(USAGE)
+  process.exit(1)
+}
+
+if (minerId && !minerId.startsWith('f0')) {
+  console.warn('Warning: miner id %s does not start with "f0", is it a valid miner address?', minerId)
+}
+
+console.error('Fetching measurements committed between %j and %j', start, end)
+const client = new pg.Client({ connectionString: DATABASE_URL })
+await client.connect()
+
+const { rows } = await client.query(`
+  SELECT cid, meridian_round FROM commitments
+  WHERE published_at >= $1 AND published_at < $2
+  `, [
+  start,
+  end
+])
+
+console.error('Found %s commitments', rows.length)
+
+// Group events by rounds
+
+/** @type {{roundIndex: bigint, measurementCids: string[]}[]} */
+const rounds = []
+for (const { meridian_round: roundIndex, cid } of rows) {
+  if (!rounds.length || rounds[rounds.length - 1].roundIndex !== roundIndex) {
+    rounds.push({ roundIndex, measurementCids: [] })
+  }
+  rounds[rounds.length - 1].measurementCids.push(cid)
+}
+
+// Discard the first and the last round, because most likely we don't have all events for them
+rounds.shift()
+rounds.pop()
+
+console.error(' → found %s complete rounds', rounds.length)
+
+const ALL_MEASUREMENTS_FILE = 'measurements-all.ndjson'
+const MINER_DATA_FILE = `measurements-${minerId}.ndjson`
+
+const allMeasurementsWriter = fs.createWriteStream(ALL_MEASUREMENTS_FILE)
+const minerDataWriter = minerId
+  ? fs.createWriteStream(MINER_DATA_FILE)
+  : null
+
+const abortController = new AbortController()
+const signal = abortController.signal
+process.on('SIGINT', () => abortController.abort(new Error('interrupted')))
+
+const resultCounts = {
+  total: 0
+}
+
+try {
+  for (const { roundIndex, measurementCids } of rounds) {
+    signal.throwIfAborted()
+    await processRound(roundIndex, measurementCids, resultCounts)
+  }
+} catch (err) {
+  if (!signal.aborted) {
+    throw err
+  }
+}
```

Comment:
IIRC, we wanted to decouple "fetch measurements" from "evaluate fetched measurements".

What do you think about modifying this script to produce output suitable for consumption by `bin/evaluate-measurements` (introduced by #401)?

Reply:
Oh, so the "fetch" part is responsible for preprocessing? I was expecting the fetch part to only fetch the measurements. I need to think about this a bit more.

Reply:
Will address this comment next week